### PR TITLE
Adds fine tuning contrib example

### PR DESCRIPTION
**Pull request by skrawcz**
This module shows how one can perform fine tuning using Hamilton. It is a basic example using the transformers library that connects to huggingface to pull and fine tune a FLAN model.

One should be able to adapt this code to their needs.

## For new dataflows:
Do you have the following?
- [x] Added a directory mapping to my github user name in the contrib/hamilton/contrib/user directory.
  - [x] If my author names contains hyphens I have replaced them with underscores.
  - [x] If my author name starts with a number, I have prefixed it with an underscore.
  - [x] If your author name is a python reserved keyword. Reach out to the maintainers for help.
  - [x] Added an author.md file under my username directory and is filled out.
  - [x] Added an __init__.py file under my username directory.
- [x] Added a new folder for my dataflow under my username directory.
  - [x] Added a README.md file under my dataflow directory that follows the standard headings and is filled out.
  - [x] Added a __init__.py file under my dataflow directory that contains the Hamilton code.
  - [x] Added a requirements.txt under my dataflow directory that contains the required packages outside of Hamilton.
  - [x] Added tags.json under my dataflow directory to curate my dataflow.
  - [x] Added valid_configs.jsonl under my dataflow directory to specify the valid configurations.
  - [x] Added a dag.png that shows one possible configuration of my dataflow.
- [x] I hearby acknowledge that to the best of my ability, that the code I have contributed contains correct attribution
and notices as appropriate.

## How I tested this
Ran it locally in a docker container

## Notes
We might want to invest in making it so that multiple modules could make a contribution because that could help segment the code better.

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Dataflow documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+]
+```
+
+You would then pass in as _inputs_ to execution `"data_path"=PATH_TO_THIS_FILE` as well as `"input_text_key"="question"` and `"output_text_key"="reply"`.
+- Instantiate the driver. Use `{"start": "base"}` as configuration to run with to use a raw base LLM to finetune.
+- Pick your LLM. `model_id_tokenizer="google/mt5-small"` is the default, but you can change it to any of the models
+that the transformers library supports for `AutoModelForSeq2SeqLM` models.
+- Run the code.
+
+Because there's no configuration that changes the shape of the DAG, you can run the code like this:
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md:66]: Clarify "Becuase there's no configuration that changes the shape of the DAG" (remove)

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# Purpose of this module
+This module shows you how to fine-tune an LLM model. This code is inspired by this [fine-tuning code](https://github.com/dagster-io/dagster_llm_finetune/tree/main).
+
+Specifically the code here, from an approach standpoint, shows Supervised Fine-Tuning (SFT) for dialogue. This approach instructs the model to be more
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md:4]: Remove "from an approach standpoint"

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+```bash
+docker build -t YOUR_IMAGE_NAME .
+docker run YOUR_IMAGE_NAME
+```
+
+# Configuration Options
+ - `{"start": "base"}` Suggested configuration to run with to use a raw base LLM to finetune.
+ - `{"start":  "presaved"}` Use this if you want to load an already fine-tuned model and then just eval it.
+
+# Limitations
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md:124]: Scaling up -- we want to point them in the right direction in case they want to use a bigger model/whatnot. Also limitation is that we have some pieces (E.G. modal support) in the works.

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+        )
+    format_data = pd.DataFrame(format_data)
+    train, test = train_test_split(format_data, random_state=random_state, train_size=train_size)
+    # split the test set into a validation set and inference set
+    validation, inference = train_test_split(
+        test, random_state=random_state, train_size=validation_size
+    )
+    dataset_train = Dataset.from_pandas(train)
+    dataset_validation = Dataset.from_pandas(validation)
+    dataset_inference = Dataset.from_pandas(inference)
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:77]: Why an inference dataset? Feels to me that that's a different pipeline -- E.G. an inference pipeline. You'll want to train the model then infer on a dataset. That'll make more sense and make it reusable. 

If we remove this we get more training/test data, and we can just add a quick repl to load the model we saved to make an inference request.

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+    return tokenizer
+
+
+def tokenized_inputs(
+    train_set: Dataset,
+    validation_set: Dataset,
+    inference_set: Dataset,
+    tokenizer: PreTrainedTokenizerBase,
+) -> DatasetType:
+    """Tokenizes"""
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:103]: Finish sentence?

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+            max_source_lengths=max_source_lengths,
+            max_target_lengths=max_target_lengths,
+            tokenizer=tokenizer,
+        ),
+        batched=True,
+        remove_columns=["input_text", "output_text"],
+    )
+
+
+def tokenized_inference(
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:218]: Having these all the same is a parallel pipeline -- maybe best to first tokenize and then split into datasets. That way part of this is reusable for inference if you want to infer in batch, and we don't repeat the same exact things three times. I'm assuming its OK to do this as its a pure map operation.

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+This module shows you how to fine-tune an LLM model. This code is inspired by this [fine-tuning code](https://github.com/dagster-io/dagster_llm_finetune/tree/main).
+
+Specifically the code here, from an approach standpoint, shows Supervised Fine-Tuning (SFT) for dialogue. This approach instructs the model to be more
+useful to directly respond to a question, rather than optimizing over an entire dialogue. SFT is the most common type of fine-tuning,
+as the other two options: Pre-training for Completion, and RLHF, required more to work. Pre-training requires more computational power,
+while RLHF requires higher-quality dialogue data.
+
+This code should work on a regular CPU (in a docker container), which will allow you to test out the code locally without
+any additional setup. This specific approach this code uses is [LoRA](https://arxiv.org/abs/2106.09685) (low-rank adaptation of large language models), which
+means that only a subset of the LLM's parameters are tweaked and prevents over-fitting.
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md:11]: Add more info about the issues with mac/MPS

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+def trainer(
+    base_model: torch.nn.Module,
+    training_args: Seq2SeqTrainingArguments,
+    data_collator: DataCollatorForSeq2Seq,
+    tokenized_train: Dataset,
+    tokenized_validation: Dataset,
+) -> Seq2SeqTrainer:
+    """The trainer we'll use for fine-tuning.
+
+    :param base_model:
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:379]: Remove params or add info about them

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+
+def metric() -> evaluate.EvaluationModule:
+    """Return the metric to use."""
+    return evaluate.load("rouge")
+
+
+def _evaluate_peft_model(sample, model, tokenizer, device, max_target_length=512) -> tuple:
+    """Helper function to evaluate the model on a sample.
+
+    :param sample:
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:475]: Same as above

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+
+
+def training_and_validation_set_metrics(
+    validation_set_metrics: dict, fit_trainer_eval_metrics: dict
+) -> dict:
+    """Combines the training and validation set metrics."""
+    return {**fit_trainer_eval_metrics, **validation_set_metrics}
+
+
+def inference_set_predictions(
```
Comment (elijahbenizzy) [inline contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:564]: This should really be a second pipeline or an external script?

**Review by elijahbenizzy — COMMENTED** [review]
I think we can simplify a bit by:
1. Removing inference from this -- that should be a snippet of code in the README of an `ipynb` or something
2. Processing features prior to splitting the dataset
3. Fixing up/removing some of the docstrings

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md (hunk 3 above)
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md:124]: that's not a limitation of this code.

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py (hunk 4 above)
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:77]: because it's a hold out set. If you're going to eval performance it should be on an unseen set.

Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:77]: can rename to hold out set.

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md (hunk 7 above)
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md:11]: I don't have much to say here though.

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py (hunk 10 above)
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:564]: no. this is hamilton -- run what you want.

Comment (elijahbenizzy) [reply contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py]: Disagree, I think this should be a different one. This does too much.

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py (hunk 4 above)
Comment (elijahbenizzy) [reply contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:77]: Yes, this relates to my whole point about too many things in this pipeline. But yeah, a hold-out set is fine.

**Review by elijahbenizzy — COMMENTED** [review]
Can we tokenize the dataset before splitting? That'll kill a lot of duplicate code and make it easier to read.

Otherwise this looks fine, I think you should break it out cause it's doing too much and I find that hard to follow, but let's just get it out.

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py (hunk 6 above)
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py]: the use case is different as the second tokenization is applied differently.

File: contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py (hunk 4 above)
Comment (skrawcz) [reply contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py:77]: Remember Hamilton allows you to choose what you compute so extra stuff isn't an issue.

In terms of code organization we should consider extending the hub to allow multiple modules in one download, since I agree this could be split up into multiple modules, but in terms of a full example I think this is the best way for now.

**Review by elijahbenizzy — COMMENTED** [review]
Looks fine, maybe `_preprocess_function` should do the dataset `.map` call

**Review by elijahbenizzy — APPROVED** [review]
